### ISSUE TITLE
Large triangles result in wrong collisions

### ISSUE THREAD

**Issue by forestrf** [forestrf/jitterphysics#3]
Large triangles result in wrong collisions.
This can be seen by executing the demo and throwing some primitives in the TriangleMeshCharaterController scene. The throwed objects sometimes start jumping, even when using spherical expansion on the mesh.

**Comment by forestrf** [comment]
After testing with concave compound shapes, the problem lies with in big concave compound shapes generating wrong contacts on XenoCollide.cs>Detect

**Comment by forestrf** [comment]
The reported point by XenoCollide.Detect is correct, but the normal and penetration isn't

**Comment by forestrf** [comment]
Partially replaced MPR from Jitter with the implementation of BEPU. Has the same error with the normal and penetration being incorrect. Tested by including the projects BEPUphysics and BEPUutilities and inserting this code inside XenoCollide.cs

```csharp

        public static bool DetectBepu(ISupportMappable support1, ISupportMappable support2, ref JMatrix orientation1,
             ref JMatrix orientation2, ref JVector position1, ref JVector position2,
             out JVector point,
             /*out JVector point1, out JVector point2,*/
             out JVector normal, out float penetration) {
            var bb1 = support1 as Shapes.BoxShape;
            var bb2 = support2 as Shapes.BoxShape;
            if (bb1 == null || bb2 == null) {
                normal = point = default;
                penetration = default;
                return false;
            }

            JQuaternion q1 = JQuaternion.CreateFromMatrix(orientation1);
            JQuaternion q2 = JQuaternion.CreateFromMatrix(orientation2);

            ConvexShape c1 = new BEPUphysics.CollisionShapes.ConvexShapes.BoxShape(bb1.Size.X, bb1.Size.Y, bb1.Size.Z);
            ConvexShape c2 = new BEPUphysics.CollisionShapes.ConvexShapes.BoxShape(bb2.Size.X, bb2.Size.Y, bb2.Size.Z);
            RigidTransform r1 = new RigidTransform(
                new Vector3(position1.X, position1.Y, position1.Z),
                new Quaternion(q1.X, q1.Y, q1.Z, q1.W)
                );
            RigidTransform r2 = new RigidTransform(
                new Vector3(position2.X, position2.Y, position2.Z),
                new Quaternion(q2.X, q2.Y, q2.Z, q2.W)
                );

            Vector3 randomPenetration = (r2.Position - r1.Position);

            bool yep = MPRToolbox.GetContact(
                c1, c2, ref r1, ref r2,
                ref randomPenetration, out var contact);

            point = new JVector(contact.Position.X, contact.Position.Y, contact.Position.Z);
            penetration = contact.PenetrationDepth;
            normal = -new JVector(contact.Normal.X, contact.Normal.Y, contact.Normal.Z);

            return yep;
        }
```
Still, collisions not always work after this replacement.

**Comment by forestrf** [comment]
While the point of collision is correct, normal and penetration isn't. That is because MPR will not output the smallest penetration depth between two convex, rather having it was a fortunate side effect that nearly always happens. Fortunately it seems that it can be obtained with some extra work http://movement.stanford.edu/courses/cs468-01-fall/Papers/van-den-bergen.pdf

**Comment by forestrf** [comment]
Solution (costly):
https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=9&t=591
```
use either EPA (expanding polytope algorithm) or sample the pentration depth in several directions, using the support mapping.
```
Those random directions could be the vertex, face normals, diff between center of shapes and others

**Comment by forestrf** [comment]
Solution: Use SAT for some, or all, collisions.

**Comment by forestrf** [comment]
Implementing SAT for fixing this issue. As a side effect, any combination of collisions between spheres and capsules got faster, but cones and cylinders may be dropped.

**Comment by forestrf** [comment]
A nearly completed implementation of SAT solved this, but it is about 3x times slower in the narrow phase por convex-convex collisions (but sphere-sphere, sphere-capsule and capsule-capsule are now about 3x times faster in the narrow phase).

**Comment by forestrf** [comment]
Fixed in 5752b084f8a1b3af43ae59290732a33f21f33123 by using SAT (capsule, cone and cylinder not done yet)

**Comment by forestrf** [comment]
Fixed MPR for oblong shapes in dd4c17c

The idea is:

1. Run MPR once and get a collision point. Check how oblong are the shapes by checking the ratio between the smallest and largest axis of the minimal bounding box of each shape. I consider a shape oblong if largestAxis/smallestAxis>2. If both shapes aren't oblong, end here.
2. Using the minimal bounding box of the shapes , get the smallest axis of that box and "resize" the box, using the collision point from the first MPR run as the resizing origin, so that all axis are equal to that smallest axis.
3. From that resized shape, obtain its center and run MPR again with the original shapes, but with this new centers (that will be inside the original shapes). End here.

This time MPR has a better hint of where the normal should point at, and after testing this I wasn't able to obtain a wrong normal.
